src/utils/actions.js

```diff
@@ -0,0 +1,321 @@
+// Context:
+//  Patient: identifier for the patient whose data triggered the workflow
+//  records: list of records
+//  Encounter: encounter where some data triggered the workflow
+//  Organization: Organization that is doing the workflow
+//  PlanDef: *** do we want to keep a "pre-processed" version of the plandef?
+//  Action: action currently in context from the planDef
+//  client: set of clients to connect to
+//      source: FHIR client to make additional queries to EHR
+//      dest: FHIR client to send to PHA/TTP
+//      trust services
+//      database
+//  reportingBundle: the entire bundle to be submitted
+//  contentBundle: just the content section of the report bundle
+//  flags: various boolean or enum flags about the status of the report.
+//         is it encrypted y/n, is it identifiable or de/pseudo/anonymized, etc
+//  cancelToken: indicates the job should be cancelled (from a process external
+//         to the job itself -- this tells the job to cease processing)
+//  exitStatus: if set, terminates the job and marks it with the given status
+//         (for example, if a given patient doesn't meet reporting criteria, don't keep reporting)
+
+// we assume the context updates and is passed to functions
+// in turn.  I'm also assuming the context is unique
+// to the current action, since we'd be cycling through actions
+// anyway.
+const Fhir = require('fhir').Fhir;
+const FhirPath = require('fhir').FhirPath;
+const fhir = new Fhir();
+const baseIgActions = {
+  'check-trigger-codes': context => {
+    const action = context.action;
+    const resources = context.records;
+
+    const boolMap = action.action[0].condition.map(condition => {
+      if (condition.expression) {
+        const expression = condition.expression;
+        // check the records for a trigger code
+        return resources.some(resource => {
+          //might have to check that this is non-empty here
+          return evaluateExpression(expression, resource);
+        });
+      } else {
+        // default true for non-applicability conditions
+        return true;
+      }
+    });
+    // boolMap should be all-true (all conditions met)
+    const triggered = boolMap.every(entry => {
+      return entry === true;
+    });
+    context.flags['triggered'] = triggered;
+  },
+  'check-participant-registration': context => {
+    // TODO: how do we check the participant registration?
+    // Through the source client?
+    return context;
+  },
+  'create-report': context => {
+    // the output might define a profile
+    let contentBundle = createBundle(context.records, 'collection');
+    if (context.action.output) {
+      const output = context.action.output[0];
+      if (output.profile) contentBundle = validateProfile(contentBundle, output.profile[0]);
+    }
+    const header = makeHeader(context);
+    context.contentBundle = contentBundle;
+    context.reportingBundle = createBundle([header, ...contentBundle.entry], 'message');
```

Comment:
The reporting bundle is supposed to contain the actual content bundle as a Bundle. So a Bundle inside a Bundle: https://build.fhir.org/ig/HL7/fhir-medmorph/StructureDefinition-us-ph-reporting-bundle.html#introduction

Also both of these bundles should have a profile on them, maybe that should be another arg to the create function

```suggestion
    context.reportingBundle = createBundle([header, contentBundle], 'message');
```

Reply:
I think you can tell based on the type right? Or perhaps instead of passing in 'message' vs 'collection' the third parameter is 'reporting' vs 'content' and that can be used to set the Bundle.type and the profile